File: utilities/dates/python/basic_utils.py
from datetime import datetime, timezone


def parse_date(value):
    if not isinstance(value, str) or not value.strip():
        return None
    try:
        normalized = value.strip().replace("Z", "+00:00")
        parsed = datetime.fromisoformat(normalized)
        return parsed if parsed.tzinfo else parsed.replace(tzinfo=timezone.utc)
    except ValueError:
        return None


def days_between(start, end):
    start_date = parse_date(start)
    end_date = parse_date(end)
    if start_date is None or end_date is None:
        return None
    return round((end_date - start_date).total_seconds() / 86400)

File: utilities/dates/javascript/basic_utils.js
export const parseDate = (value) => {
    if (typeof value !== 'string' || value.trim() === '') return undefined;
    const date = new Date(value);
    return Number.isNaN(date.getTime()) ? undefined : date;
};

export const daysBetween = (start, end) => {
    const startDate = parseDate(start);
    const endDate = parseDate(end);
    if (!startDate || !endDate) return undefined;
    return Math.round((endDate.getTime() - startDate.getTime()) / 86400000);
};
